### PR TITLE
revamp startup and shutdown flow

### PR DESCRIPTION
**Pull request by WindzCUHK**
# Description

- revamp startup flow
  - retry
    - allow cancel during retry
    - mode=init, max. retry time = `3 min`
  - control New() phase by `initCtx`
    - On error, fatal
    - On signal, finish current function block to ensure integrity, skip further function block, handle context
  - control Start() phase by `runCtx`
    - On error in main go routine, cancel context
    - On error in sub go routine, should continue except for fatal error
    - On signal, finish current function block to ensure integrity, skip further function block, handle context
    - On context cancel, should stop all sub go routine gracefully, and skip further process or retry
- revamp shutdown flow
  - Once Start(), should always Shutdown()
  - Shutdown() is graceful and blocking

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code

## Related issue/PR

- Related #95
- Fixes #98
- Fixes #101
- Created #108

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [ ] Passed all pipeline checking

## Checklist for maintainer

- Use `Squash and merge`
- Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- Delete the branch after merge

### CODE REVIEW

File: cmd/athenz-sia/main.go [hunk 1]
```diff
@@ -73,41 +79,93 @@ func main() {
...
-	log.Infof("Booting up with args: %v, config: %+v", os.Args, idConfig)
+	log.Infof("Booting up with args: %v, config: %+v", os.Args, idCfg)
 
-	certificateChan := make(chan struct{}, 1)
-	ch := make(chan os.Signal, 1)
-	signal.Notify(ch, syscall.SIGTERM, os.Interrupt)
+	// delay boot with jitter
+	if idCfg.DelayJitterSeconds != 0 {
+		sleep := time.Duration(rand.Int63n(idCfg.DelayJitterSeconds)) * time.Second
+		log.Infof("Delaying boot with jitter [%s] randomized from [%s]...", sleep, time.Duration(idCfg.DelayJitterSeconds)*time.Second)
```
Comment (ctyano) [inline cmd/athenz-sia/main.go:89]: @WindzCUHK Several log printings are added outside the purpose of "revamp startup and shutdown flow".
I suppose the log printing mechanism and the messages should be well discussed before merging this PR.
Could you separate this mechanism to a different PR and have discussion there?

**Comment by ctyano** [comment]
> Description
> 
>     revamp startup flow
>     revamp shutdown flow

Regardless of this description, I see a lot of changes and I think this description itself needs to be updated to explain it in more detail.

File: cmd/athenz-sia/main.go [hunk 2]
```diff
@@ -73,41 +79,93 @@ func main() {
...
+		time.Sleep(sleep)
+	}
+
+	// register metrics
+	metrics.RegisterBuildInfo(filepath.Base(os.Args[0]), VERSION, BUILD_DATE)
+
+	// variables
+	causeBySignal := fmt.Errorf("received signal")
```
Comment (ctyano) [inline cmd/athenz-sia/main.go:97]: The signal messages should be well discussed before merging this PR as well.

File: cmd/athenz-sia/main.go [hunk 3]
```diff
@@ -75,41 +75,93 @@ func main() {
...
+	}()
 
-	err, sdChan := identity.Certificated(idConfig, certificateChan)
+	// initiate background services
+	certService, err := certificate.New(initCtx, idCfg)
 	if err != nil {
-		log.Fatalln(err)
+		log.Fatalf("Error initiating certificate provider: %s", err.Error())
+	}
+	tokenService, err := token.New(initCtx, idCfg)
```
Comment (ctyano) [inline cmd/athenz-sia/main.go:114]: @WindzCUHK Is there any particular reason of changing the naming style from using the word "daemon" to using the word "service"?
I see the usage is quite similar and if the purpose is to just inherit the naming style that existed in Athenz/athenz-client-sidecar, then I don't believe the reason is sufficient to break the consistency.

File: pkg/certificate/daemon.go [hunk 4]
```diff
@@ -12,41 +12,51 @@
...
+
+func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, error) {
+	if ctx.Err() != nil {
+		log.Info("Skipped certificate provider initiation")
+		return nil, nil
 	}
 
-	if idConfig.ProviderService == "" {
-		log.Infof("Certificate provisioning is disabled with empty options: provider service[%s]", idConfig.ProviderService)
+	if idCfg.ProviderService == "" {
```
Comment (ctyano) [inline pkg/certificate/daemon.go:48]: @WindzCUHK I also see that this change is unnecessary.

File: pkg/metrics/daemon.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	return nil
+}
+
+func (ms *metricsService) Shutdown() {
+	log.Info("Initiating shutdown of metrics exporter daemon ...")
+	close(ms.shutdownChan)
+
+	if ms.exporter != nil && ms.exporterRunning {
+		err := ms.exporter.Shutdown() // context.Background() is used, no timeout
+		// P.S. Make sure to use the httpChecker to ensure ListenAndServe() is finished before Shutdown() is called. If ListenAndServe() does not finish creating the server object before Shutdown() is called, the internal server field will be nil and Shutdown() be a no-op. ListenAndServe() will block and cause deadlock.
```
Comment (ctyano) [inline pkg/metrics/daemon.go:134]: @WindzCUHK Did you write this message by yourself or just brought it from some where else?
If the message is just brought it from some where else, I suppose we should at least leave the link to the original message.

File: cmd/athenz-sia/main.go (hunk 3 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go:114]: It is my personal understand. Please correct me.
- daemon implies something running in background.
- service provide some feature for the end user.
- From the user's point of view, the token service provides token as a service.
- From the developer's point of view, the token service is able to run in background so it should implements the daemon interface.
- The main focus should be providing some services to the end user. Hence, it should be service rather than daemon. (if it should be called as daemon service, simplify as service for the name.)

File: cmd/athenz-sia/main.go [hunk 6]
```diff
@@ -15,35 +15,34 @@
...
-var (
-	APP_NAME   = version.APP_NAME
-	VERSION    = version.VERSION
-	BUILD_DATE = version.BUILD_DATE
-)
-
 // printVersion returns the version and the built date of the executable itself
 func printVersion() {
-	if VERSION == "" || BUILD_DATE == "" {
+	if version.VERSION == "" || version.BUILD_DATE == "" {
```
Comment (ctyano) [inline cmd/athenz-sia/main.go:41]: @WindzCUHK Maybe this is a little relevant from this PR's context and should have been pointed in https://github.com/AthenZ/k8s-athenz-sia/pull/105/files but, but the building environment parameter is not only for version management and the perspective should not be narrowed for only version management, thus I suppose the struct itself should be treated as a more generic usage.(e.g. arguments.go)

File: cmd/athenz-sia/main.go (hunk 2 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go:93]: extra review

File: cmd/athenz-sia/main.go (hunk 3 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go:114]: extra review

File: pkg/metrics/daemon.go (hunk 5 above)
Comment (WindzCUHK) [reply pkg/metrics/daemon.go:134]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/100/commits/ecabacf69caee7f455effe795600728ea71fb361

File: pkg/certificate/daemon.go (hunk 4 above)
Comment (WindzCUHK) [reply pkg/certificate/daemon.go]: - revert in commit: https://github.com/AthenZ/k8s-athenz-sia/pull/100/commits/adf7ade47db7b8736c3ad7bd49020c181680e05d
- split as PR: https://github.com/AthenZ/k8s-athenz-sia/pull/108

File: cmd/athenz-sia/main.go (hunk 6 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go:41]: refactor in different PR?

- `version.go` (appName, version, etc.): https://github.com/AthenZ/k8s-athenz-sia/pull/105
- `cache.go` (pod, ns, etc.): https://github.com/AthenZ/k8s-athenz-sia/pull/103

File: cmd/athenz-sia/main.go (hunk 3 above)
Comment (ctyano) [reply cmd/athenz-sia/main.go]: @WindzCUHK Just to clarify, but I'm not actually against this change. I'm only concerning about the consistency.

File: pkg/certificate/daemon.go [hunk 7]
```diff
@@ -283,105 +279,97 @@ func Certificated(idConfig *config.IdentityConfig, stopChan <-chan struct{}) (er
...
+		if err != nil {
+			log.Warnf("Failed to delete x509 certificate Instance ID record: %s", err.Error())
+		} else {
+			log.Infof("Successfully deleted x509 certificate Instance ID record[%s]", cs.handler.InstanceID())
+		}
+	}
+}
+
+// newExponentialBackOff returns a backoff config with first retry delay of 5s. Allow cancel by context.
+func newExponentialBackOff(ctx context.Context, maxElapsedTime time.Duration) backoff.BackOff {
```
Comment (ctyano) [inline pkg/certificate/daemon.go:368]: @WindzCUHK I don't believe it's necessary to change this function name from `getExponentialBackoff` since we're just replacing it.

File: cmd/athenz-sia/main.go (hunk 1 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go]: - log message content
- log level
- is current log message enough?

File: pkg/certificate/daemon.go (hunk 7 above)
Comment (WindzCUHK) [reply pkg/certificate/daemon.go:368]: new() creates an object. get() get an existing object.
Using new() as function name is more suitable.

File: cmd/athenz-sia/main.go (hunk 6 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go:41]: fix in PR: https://github.com/AthenZ/k8s-athenz-sia/pull/108/commits/a04d5afd0a834717b01aec3bd94dbb90f9c2432c

File: pkg/config/default.go [hunk 8]
```diff
@@ -47,6 +47,9 @@ var (
 	// default values for graceful shutdown
 	DEFAULT_SHUTDOWN_TIMEOUT = 5 * time.Second
 	DEFAULT_SHUTDOWN_DELAY   = time.Duration(0)
+
+	// default maximum elapsed time on initialization
+	DEFAULT_MAX_ELAPSED_TIME_ON_INIT = 3 * time.Minute
```
Comment (ssunotz) [inline pkg/config/default.go:52]:
```suggestion
	DEFAULT_MAX_ELAPSED_TIME_ON_INIT = 1 * time.Minute
```

File: cmd/athenz-sia/main.go (hunk 1 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go]: reviewed

File: cmd/athenz-sia/main.go (hunk 3 above)
Comment (WindzCUHK) [reply cmd/athenz-sia/main.go]: filename: `daemon.go` => `service.go`

File: pkg/certificate/identity.go [hunk 9]
```diff
@@ -256,6 +256,7 @@ func (h *identityHandler) GetX509RoleCert() (rolecerts [](*RoleCertificate), rol
 		t.TLSClientConfig.RootCAs = certPool
 	}
 
+	// TODO: no need to renew ZTS Client after https://github.com/AthenZ/k8s-athenz-sia/pull/99
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply pkg/certificate/identity.go:259]: https://github.com/AthenZ/k8s-athenz-sia/issues/109

**Review by ssunotz — APPROVED** [review]
